build(cmake): Allow ystdlib-cpp to be imported into other CMake projects as a library (#65)
y-scope/ystdlib-cpp: CMakeLists.txt

> cmake_minimum_required(VERSION 3.23)

Reviewer: This is breaking change for things that depend on ystdlib (e.g., CLP still supports Ubuntu Jammy, so it'll now need to install CMake 3.23), so we should indicate as much in the title.

> project(ystdlib VERSION "0.0.1" LANGUAGES CXX)

Reviewer: ```suggestion
project(ystdlib VERSION "0.1.0" LANGUAGES CXX)
```

For new projects, we start our versioning at 0.1.0.

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")

> include(ystdlib-helpers)
> include(CMakePackageConfigHelpers)
> include(GNUInstallDirs)

Reviewer: ```suggestion
include(CMakePackageConfigHelpers)
include(GNUInstallDirs)
include(ystdlib-helpers)
```

To alphabetize.

option(ystdlib_BUILD_TESTING "Build the testing tree for ystdlib." ON)

# Require compiler versions that support the C++20 features necessary for compiling ystdlib
if("AppleClang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(ystdlib_CMAKE_CXX_COMPILER_MIN_VERSION "16")
elseif("Clang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(ystdlib_CMAKE_CXX_COMPILER_MIN_VERSION "16")
elseif("GNU" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(ystdlib_CMAKE_CXX_COMPILER_MIN_VERSION "11")
else()
    message(
        FATAL_ERROR
        "Unsupported compiler: ${CMAKE_CXX_COMPILER_ID}. Please use AppleClang, Clang, or GNU."
    )
endif()
if("${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "${ystdlib_CMAKE_CXX_COMPILER_MIN_VERSION}")
    message(
        FATAL_ERROR
        "${CMAKE_CXX_COMPILER_ID} version ${CMAKE_CXX_COMPILER_VERSION} is too low. Must be at \
        least ${ystdlib_CMAKE_CXX_COMPILER_MIN_VERSION}."
    )
endif()

set(CMAKE_EXPORT_COMPILE_COMMANDS
    ON
    CACHE BOOL
    "Enable/Disable output of compile commands during generation."
    FORCE
)

set(ystdlib_LIBRARIES
    "containers"
    "error_handling"
    "io_interface"
    "wrapped_facade_headers"
    CACHE STRING

>     "Semi-colon separated list of libraries to be built."

Reviewer: ```suggestion
    "Semicolon-separated list of libraries to be built."
```

)

message(STATUS "Building the following libraries:")
foreach(LIB IN LISTS ystdlib_LIBRARIES)

>     message(" - ${LIB}")

Reviewer: ```suggestion
    message(STATUS " - ${LIB}")
```

Unless we purposely didn't include `STATUS`?

    catch_discover_tests(${UNIFIED_UNIT_TEST_TARGET} WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/testbin)
endif()

set(ystdlib_INSTALL_CONFIG_DIR ${CMAKE_INSTALL_LIBDIR}/cmake/ystdlib)

> set(ystdlib_INSTALL_INCLUDE_DIR ${CMAKE_INSTALL_INCLUDEDIR})

Author: I believe we are no longer using project prefix uppercase style.

    catch_discover_tests(${UNIFIED_UNIT_TEST_TARGET} WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/testbin)
endif()

> set(ystdlib_INSTALL_CONFIG_DIR ${CMAKE_INSTALL_LIBDIR}/cmake/ystdlib)

Reviewer: Move after the `add_subdir` call so it's closer to its use?

add_subdirectory(src/ystdlib)

configure_package_config_file(
    ${CMAKE_CURRENT_LIST_DIR}/cmake/ystdlib-config.cmake.in

>     ${CMAKE_CURRENT_BINARY_DIR}/ystdlib-config.cmake

Reviewer: Could extract this into a variable since it has multiple uses?

    INSTALL_DESTINATION ${ystdlib_INSTALL_CONFIG_DIR}
)

write_basic_package_version_file(

>     ${CMAKE_CURRENT_BINARY_DIR}/ystdlib-config-version.cmake

Reviewer: Could extract this into a variable since it has multiple uses?

    catch_discover_tests(${UNIFIED_UNIT_TEST_TARGET} WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/testbin)
endif()

# We require all libraries use the same minimum version of dependencies to avoid issues.
set(BOOST_MIN_VERSION "1.81.0")

set(CONFIG_PATH_SUFFIX "cmake/ystdlib")
set(CONFIG_LIBS_PATH_SUFFIX "${CONFIG_PATH_SUFFIX}/libs")

# Used by libraries and must come before add_subdirectory.
set(CONFIG_LIBS_DEST_DIR "${CMAKE_INSTALL_LIBDIR}/${CONFIG_LIBS_PATH_SUFFIX}")
set(CONFIG_LIBS_INPUT_DIR "${PROJECT_SOURCE_DIR}/${CONFIG_LIBS_PATH_SUFFIX}")

add_subdirectory(src/ystdlib)

set(CONFIG_FILE_PREFIX "ystdlib-config")

> set(CONFIG_INSTALL_DIR ${CMAKE_INSTALL_LIBDIR}/${CONFIG_PATH_SUFFIX})

Reviewer: ```suggestion
set(CONFIG_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}/${CONFIG_PATH_SUFFIX}")
```

    catch_discover_tests(${UNIFIED_UNIT_TEST_TARGET} WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/testbin)
endif()

set(YSTDLIB_INSTALL_CONFIG_DIR ${CMAKE_INSTALL_LIBDIR}/cmake/ystdlib)
set(YSTDLIB_INSTALL_INCLUDE_DIR ${CMAKE_INSTALL_INCLUDEDIR})

file(GLOB_RECURSE SOURCE_FILES
  "${PROJECT_SOURCE_DIR}/src/*.cpp"
  "${PROJECT_SOURCE_DIR}/src/*.hpp"
  "${PROJECT_SOURCE_DIR}/src/*.h"
)
list(FILTER SOURCE_FILES EXCLUDE REGEX ".*test_*\..*")

add_library(ystdlib ${SOURCE_FILES})
add_library(ystdlib::ystdlib ALIAS ystdlib)

# target_link_libraries(ystdlib
#     PUBLIC
#     outcome

> # )

Author: I think this is fixed?
